### ISSUE TITLE
Transforming features by scaling

### ISSUE THREAD

**Issue by nsankar** [oracle/tribuo#242]
**System details**
- Tribuo version 4.2.1  , 
- Java version 10+
- OS/Architecture Windows / Linux

How can we transform features by scaling each feature to a given range like 0 to 1 in Tribuo? For instance in the HDBSCAN after creating the MutableDataset<>(csvSource) from the CSV.

Is there a built in Tribuo function that can be applied on the mutableDataset for scaling the data from 0 to 1 range?

**Comment by Craigacp** [comment]
Yep, we cover this in the configuration tutorial here - https://tribuo.org/learn/4.2/tutorials/configuration-tribuo-v4.html#Feature-Transformations

**Comment by nsankar** [comment]
@Craigacp Thanks for the info.

Using the transformation (0 to 1) cited above, I created a transformed model for the HDBSCAN. However, I could not invoke the get the hdbscan outlier and labels from getOutlierScores() and getClusterLabels() which are available on the hdbscan model. Can you pls, share the code snippet on how I can (a) cluster labels (b) outlier scores and (c) predictions from test datasets from a transformed hdbscan model ?

Below is the code snippet I had used for creating a hdbscan model with transformed dataset.

`NeighboursQueryFactory kdTreeFactory = new KDTreeFactory(DistanceType.L2, 4);
  HdbscanTrainer trainer = new HdbscanTrainer(20,15,kdTreeFactory); 
 //Scaling start 
  var transformations = new TransformationMap(List.of(new LinearScalingTransformation(0,1)));
  var transformed = new TransformTrainer(trainer,transformations);		
  var transformedModel = transformed.train(dataset);  //csv dataset
  // Scaling end

// This does not work on the transformedModel
 List<Integer> clusterLabels = transformedModel.getClusterLabels();
//	    var labelsSet = new HashSet<>(clusterLabels);
 List<Double> outlierScores = transformedModel.getOutlierScores();`

**Comment by Craigacp** [comment]
Looks like we were missing the accessor on `TransformedModel` that would let you see the cluster centroids. I've made a PR for it.

For the time being you can replicate the scaling and prediction code in `TransformTrainer.train` and `TransformedModel.predict` to achieve the same goals without using the accessors. Note that the cluster centroids will be in the *transformed* space, not the original feature space, and you'll need to apply the appropriate conversion yourself. Tribuo can't automatically do this as some feature transformations are not invertible.

**Comment by nsankar** [comment]
@Craigacp For point (c), The predictions output from the Transformed HDBSCAN model (with scaled data) is as below. There seems to be some issue with the transformed model and its predict . The output does not provide the predicted anomaly scores. I used the same train dataset and the test dataset (CSV) as given in the HDBSCAN tutorial

![image](https://user-images.githubusercontent.com/431890/174521997-54c4b63e-3932-4720-ab5f-3ac85ece931e.png)

**Comment by Craigacp** [comment]
Which version of the tutorial are you using? The behaviour is different in 4.3.0-SNAPSHOT (which you'll find in the tutorials directory of `main` in this repo) vs https://tribuo.org/learn/4.2/tutorials/clustering-hdbscan-tribuo-v4.html, as the behaviour of noise points has changed.

Are the assigned labels the same and it's just the anomaly scores which differ?

**Comment by nsankar** [comment]
@Craigacp  Let me check and revert.

**Comment by nsankar** [comment]
@Craigacp 

I am using the 4.3.0-Snapshot .
![image](https://user-images.githubusercontent.com/431890/174525676-c37cd61f-ef5d-4d77-ab5c-77656ca500f7.png)

I was using the earlier HDBSCAN tutorial. I now changed the HDBSCAN train code as per the one inside "main" in the repo.

The HDBSCAN train and test data I am using is as per 
https://github.com/oracle/tribuo/tree/main/tutorials/simple-2d-data

Now with the transformed data / model and its predict . i.e. as below 
`var predictions = transformedModel.predict(predictDataset);
	    System.out.println("Predictions........... : ");
	    System.out.println(predictions);`
	    
I get the following prediction output . The outlier score is a bit different for the test dataset points Vis a vis the tutorial. Perhaps it is due to the transformed (0 to 1 scaled) train and predict data ?
i.e. for test point 4.25,1.5 , the outlier score is 0.84819
for test data point (3.9,4.5), the outlier score is 0.0 and for test data point (1.5,3), outlier score is 0.013
![image](https://user-images.githubusercontent.com/431890/174527183-cd11d270-3460-4598-be1a-001b644c49ac.png)

Also if you see the prediction output above , the  OutPutScores=} is blank.
